src/main/cpp/src/cast_string.cu

```diff
@@ -48,8 +54,126 @@ constexpr bool is_whitespace(char const chr)
   }
 }
 
+template <typename T, std::enable_if_t<cuda::std::is_signed_v<T>>* = nullptr>
+T __device__ generic_abs(T value)
+{
+  return numeric::detail::abs(value);
+}
+
+template <typename T, std::enable_if_t<not cuda::std::is_signed_v<T>>* = nullptr>
+constexpr T __device__ generic_abs(T value)
+{
+  return value;
+}
+
+/**
+ * @brief determines if overflow will occur when multiplying
+ * a value by 10
+ *
+ * @tparam T type of the values
+ * @param val value to test
+ * @param adding adding or subtracting values
+ * @return true if overflow will occur by multiplying val by 10
+ */
 template <typename T>
-__global__ void string_to_integer_kernel(T* out,
+bool __device__ will_overflow(T const val, bool adding)
+{
+  if constexpr (std::is_signed_v<T>) {
+    if (!adding) {
+      auto constexpr minval = cuda::std::numeric_limits<T>::min() / 10;
+      if (val < minval) {
+        // overflow
+        return true;
+      }
+      return false;
+    }
+  }
+
+  auto constexpr maxval = cuda::std::numeric_limits<T>::max() / 10;
+  if (val > maxval) {
+    // overflow
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * @brief determines if overflow will occur when adding or subtracting values
+ *
+ * @tparam T type of the values
+ * @param lhs left hand side of the operation
+ * @param rhs right hand side of the operation
+ * @param adding true if adding, false if subtracting
+ * @return true if overflow will occur from the operation
+ */
+template <typename T>
+bool __device__ will_overflow(T const lhs, T const rhs, bool adding)
+{
+  if constexpr (std::is_signed_v<T>) {
+    if (!adding) {
+      auto const min_val = cuda::std::numeric_limits<T>::min() + rhs;
+      if (lhs < min_val) {
+        // overflow
+        return true;
+      }
+      return false;
+    }
+  }
+
+  auto const max_val = cuda::std::numeric_limits<T>::max() - rhs;
+  if (lhs > max_val) {
+    // overflow
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * @brief process a single digit of input
+ *
+ * @tparam T type of the input data
+ * @param first_value true if this is the first value added
+ * @param current_val current computed value
+ * @param new_digit digit to append to the computed value
+ * @param adding true if adding, false if subtracting
+ * @return true if success, false if overflow
+ */
+template <typename T>
+bool __device__ process_value(bool first_value, T& current_val, T const new_digit, bool adding)
```

Comment:
We should not use ref in input parameter. Instead, return a pair of `<ret, bool>`.

Reply:
Updated to fix this, but I went with `<bool, ret>` for no particular reason.